WIP: added SnakeMouseListener class

rachel613g opened the pull request.

rachel613g commented:
This doesn't actually work. I move my mouse and the snake does not move. I do not know if the method gets called at all. I am in midst of a test that I did not push because it doesn't run for lack of understanding with mocks.

rachel613g commented:
> This doesn't actually work. I move my mouse and the snake does not move. I do not know if the method gets called at all. I am in midst of a test that I did not push because it doesn't run for lack of understanding with mocks.

rachel613g commented:
There seems to be response to the movement of the mouse but it doesn't seem to be fast enough. Or maybe my logic isn't precise enough.

andrewoid commented on src/main/java/touro/snake/SnakeMouseListener.java:
```diff
@@ -0,0 +1,26 @@
+    public void mouseMoved(MouseEvent e) {
+        super.mouseMoved(e);
+        System.out.println("Mouse moved: (" + e.getX() +", " + e.getY() + ")");
+        if (e.getY() < snake.getHead().getY()) {
```
1. Don't call snake.getHead() multiple times
2. Write a comment explaining how you want this code to work.

rachel613g commented:
I updated the logical flow because before any movement other than up was being ignored because of the order of the if statements in mouseMoved() method. My logic still isn't perfect so not all mouse movements are registered when they should be. There is another significant bug that the snake will freeze and not resume moving, which did not happen to me in the same way when using the keyboard keys to move the snake so it is definitely something to do with what I added. This pull request is still a work in progress.

rachel613g commented:
I know I shouldn't usually force-push. I wanted to alter the commit message of the last commit I made but I already pushed it to the remote repository.

andrewoid commented on src/main/java/touro/snake/SnakeMouseListener.java:
```diff
@@ -0,0 +1,41 @@
+            snake.turnTo(Direction.West);
+        }
+        //mouse moved above snake head
+        if (mouseY < headY) {
```
This will work in theory, but in practice I think this is very difficult to control with the mouse because these if statements will conflict.

rachel613g replied:
I changed this code to a better variation of what I was trying to do that I found on stack exchange. 
I don't think I fully understood how the if statements I had before would have conflicted. Do you think there could have been a way I could have updated my original code with the correct logic? The original code I wanted to use can be viewed in commit 84e4202 and stack exchange logic that I use now can be found in commit 09a0184. 
The stack exchange logic is still not perfect. The stack exchange post mentioned a concept of a [line drawing algorithm ](http://en.wikipedia.org/wiki/Line_drawing_algorithm )that can be applied to this use case and that may have worked better and I hope could perform as well using the mouse works in slither.io.

rachel613g replied:
I planned on leaving the mouse logic as it is now. Is that okay?

rachel613g replied:
Hmmm, maybe AStar can be used here...

rachel613g replied:
I'm not really sure what I thought the issue was the logic I found on StackExchange works great!

andrewoid commented on src/test/java/touro/snake/SnakeMouseListenerTest.java:
```diff
@@ -0,0 +1,55 @@
+import static org.junit.Assert.*;
+import static org.mockito.Mockito.*;
+
+public class SnakeMouseListenerTest {
```
You are missing 2 tests

rachel613g replied:
added requested tests in commit [78a50b7](`https://github.com/andrewoid/TouroSnake/pull/43/commits/78a50b73df560a107cc416e1531a13102c1ce980`)

andrewoid commented on src/main/java/touro/snake/SnakeMouseListener.java:
```diff
@@ -0,0 +1,41 @@
+
+import static java.lang.Math.abs;
+
+public class SnakeMouseListener extends MouseMotionAdapter {
```
You should put in a javadoc comment explaining how this works.

rachel613g replied:
I added a comment to the code in commit https://github.com/andrewoid/TouroSnake/pull/43/commits/479ae66424ac532d62566b33e3f379864759d3f4 let me know if that is what you meant by a javadoc.

andrewoid commented on src/main/java/touro/snake/SnakeMouseListener.java:
```diff
@@ -0,0 +1,48 @@
+        //logic taken from https://bit.ly/36Gu7GG
+        //determine x and y values of the slope of line between mouse position
+        //and snake head position
+        int deltaX = e.getX() - (snake.getHead().getX() * 10);
```
use CELL_SIZE instead of 10

andrewoid replied:
apply Math.abs() here instead of in the if statements

rachel613g replied:
made these changes in commits 5728378 and 20a8fa7

andrewoid commented on src/test/java/touro/snake/SnakeMouseListenerTest.java:
```diff
@@ -0,0 +1,101 @@
+public class SnakeMouseListenerTest {
+
+    @Test
+    public void mouseMoved_left() {
```
use "east" instead of "left" (and make the same changes below)

rachel613g replied:
made this change in commit 20a8fa7
